File: labs/Module_4_Python_lab_1/Cookies.py

from heapq import heapify, heappush, heappop

def candies_sweetness_target():
    try:
        target_sweetness = int(input())
        candies = list(map(int, input().split()))
    except ValueError:
        print("Invalid input")
        return

    heapify(candies)  # Convert the list to a min-heap
    steps = 0

    while len(candies) > 1:
        # Combine the two least sweet candies
        least_sweet = heappop(candies)
        second_least_sweet = heappop(candies)
        new_sweetness = least_sweet + 2 * second_least_sweet
        heappush(candies, new_sweetness)
        steps += 1

        # Check if the new candy's sweetness exceeds the target
        if candies[0] >= target_sweetness:
            break

    print(f"{steps}")

candies_sweetness_target()

File: labs/Module_4_JAVA_LAB_1/Cookies.java
import java.util.PriorityQueue;
import java.util.Scanner;

public class Cookies {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Read the target sweetness
        int targetSweetness = scanner.nextInt();
        scanner.nextLine(); // Consume the newline

        // Read the sweetness of each candy
        PriorityQueue<Integer> candyQueue = new PriorityQueue<>();
        String[] candies = scanner.nextLine().split(" ");
        for (String candy : candies) {
            candyQueue.add(Integer.parseInt(candy));
        }

        int steps = 0;

        // Combine candies until the smallest candy in the queue meets or exceeds the target sweetness
        while (candyQueue.size() > 1 && (candyQueue.peek() < targetSweetness)) {
            int leastSweet = candyQueue.poll();
            int secondLeastSweet = candyQueue.poll();

            // Calculate the new candy sweetness
            int newCandy = leastSweet + 2 * secondLeastSweet;
            candyQueue.add(newCandy);
            steps++;
        }

        // Check if we have reached or exceeded the target sweetness with the available candies
        if (candyQueue.peek() >= targetSweetness) {
            System.out.println(steps);
        } else {
            System.out.println("It is not possible to reach or exceed the target sweetness with the given candies.");
        }

        scanner.close();
    }
}
                                                                                                                            